### PR TITLE
[Clang] Extend lifetime bound analysis to support assignments

### LINKED ISSUE #54492: Attribute clang::lifetimebound produces no warning when assign temporary's address to pointer declared before.
[[clang::lifetimebound]] could be used to provide some lifetime annotation to functions. In the statement of a pointer initialization, everything works fine. But in the statement of a pointer assignment, the desired warning can not be produced.

Here is the code sample:
```c++
#include <string>

const char* get(const std::string& s [[clang::lifetimebound]]){
    return s.data();
}

int main(){
 // can be detected
 const char* s1 = get("hello"s);

 // cannot be detected
 const char* s2 = nullptr;
 s2 = get("hello"s);

 return 0;
}

```
Here is more code to illustrate this: https://godbolt.org/z/fhTPYTa1d

Is this intentional or a flaw of lifetime check?

### PR DESCRIPTION
**Pull request by hokein**
The lifetime bound warning in Clang currently only considers initializations. This patch extends the warning to include assignments.

- **Support for assignments of built-in pointer types**: this is done is by reusing the existing statement-local implementation. Clang now warns if the pointer is assigned to a temporary object that being destoryed at the end of the full assignment expression.

With this patch, we will detect more cases under the on-by-default diagnostic `-Wdangling`. I have added a new category for this specific diagnostic so that people can temporarily disable it if their codebase is not yet clean.

This is the first step to address #63310, focusing only on pointer types. Support for C++ assignment operators will come in a follow-up patch.

Fixes #54492

### CODE REVIEW

**Comment by Xazax-hun** [comment]
Wow, this is awesome! Thanks for tackling this!

**Comment by ilya-biryukov** [comment]
@hokein could you please split this into two PRs, one with an NFC change and another one with adding assignment support?
GitHub is not great at providing the UI to review individual commits and we won't be able to merge as 2 separate commits with the green button (which seems like a desirable end state for the readability of the commit history).

**Comment by Xazax-hun** [comment]
+1, I like Ilya's suggestion.

**Comment by hokein** [comment]
I have separated the refactoring change in #96758. This PR now only focuses on the assignment support.

File: clang/lib/Sema/CheckExprLifetime.cpp [hunk 1]
```diff
@@ -964,11 +966,26 @@ static bool pathOnlyInitializesGslPointer(IndirectLocalPath &Path) {
...
+  const InitializedEntity *ExtendingEntity = nullptr;
+  if (auto IEntityP = std::get_if<const InitializedEntity *>(&CEntity)) {
+    InitEntity = *IEntityP;
+    auto LTResult = getEntityLifetime(InitEntity);
+    LK = LTResult.getInt();
+    ExtendingEntity = LTResult.getPointer();
+  } else if (auto AEntityP = std::get_if<const AssignedEntity *>(&CEntity)) {
+    AEntity = *AEntityP;
+    if (AEntity->LHS->getType()->isPointerType()) // builtin pointer type
+      LK = LK_Extended;
```
Comment (Xazax-hun) [inline clang/lib/Sema/CheckExprLifetime.cpp:985]: I am a bit confused here, could you elaborate why we want `LK_Extended` here? As fas as I remember, assignments are not doing lifetime extension.

File: clang/test/SemaCXX/attr-lifetimebound.cpp [hunk 2]
```diff
@@ -40,6 +40,12 @@ namespace usage_ok {
   int *p = A().class_member(); // expected-warning {{temporary whose address is used as value of local variable 'p' will be destroyed at the end of the full-expression}}
   int *q = A(); // expected-warning {{temporary whose address is used as value of local variable 'q' will be destroyed at the end of the full-expression}}
   int *r = A(1); // expected-warning {{temporary whose address is used as value of local variable 'r' will be destroyed at the end of the full-expression}}
+
+  void test_assignment() {
```
Comment (usx95) [inline clang/test/SemaCXX/attr-lifetimebound.cpp:44]: I found a minor case where this does not mirror initialization
```cpp
int *get(const int &x [[clang::lifetimebound]]);
void foo() {
  int *y = {get(1)}; // warning as expected.
  y = {get(1)}; // no warning.
}
```
This does not look like a commonly occurring pattern but I would be curious why this misbehaves.

File: clang/lib/Sema/CheckExprLifetime.cpp [hunk 3]
```diff
@@ -1028,6 +1045,7 @@ void checkExprLifetime(Sema &SemaRef, const InitializedEntity &Entity,
 
       switch (shouldLifetimeExtendThroughPath(Path)) {
       case PathLifetimeKind::Extend:
+        assert(InitEntity && "Expect only on initializing the entity");
```
Comment (usx95) [inline clang/lib/Sema/CheckExprLifetime.cpp:1048]: Maybe make it clearer that we do not expect lifetime extension for assignment:
 "Lifetime extension should happen only for initialization and not assigment". Same below.

**Review by usx95 — COMMENTED** [review]
You can also add "Fixes: https://github.com/llvm/llvm-project/issues/54492" to description to close this bug as this only asks for pointer type support

Please also add release notes.

**Review by hokein — COMMENTED** [review]
thanks for the review.

**Comment by hokein** [comment]
> You can also add "Fixes: #54492" to description to close this bug as this only asks for pointer type support
> 
> Please also add release notes.

Done.

**Review by Xazax-hun — APPROVED** [review]
LGTM, with a nit.

**Review by usx95 — APPROVED** [review]
Thanks. LGTM.

File: clang/lib/Sema/CheckExprLifetime.cpp [hunk 4]
```diff
@@ -964,17 +966,34 @@ static bool pathOnlyInitializesGslPointer(IndirectLocalPath &Path) {
   return false;
 }
 
-void checkExprLifetime(Sema &SemaRef, const InitializedEntity &Entity,
+void checkExprLifetime(Sema &SemaRef, const CheckingEntity &CEntity,
```
[the remark this answers is not in the document]
Comment (hokein) [reply clang/lib/Sema/CheckExprLifetime.cpp]: Thanks for the suggestion. I'm not a fan of `std::variant` either. I've added a new overload for `AssignedEntity`.

**Review by Xazax-hun — APPROVED** [review]

**Comment by zmodem** [comment]
We're seeing some -Wdangling-assignment warnings after this change, and they all seem to be false positives so far: https://crbug.com/350808950 It's not a huge deal if there are just a few, but figured it's worth mentioning.

**Comment by hokein** [comment]
> We're seeing some -Wdangling-assignment warnings after this change, and they all seem to be false positives so far: https://crbug.com/350808950 It's not a huge deal if there are just a few, but figured it's worth mentioning.

Thanks for the report. That's interesting. Technically it is a dangling assignment. The tricky bit is that the pointer being assigned is in a system header (implementation details of the system macro `A2CW`). 
Clang by default [suppresses diagnostics](https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-in-system-headers) in system headers, but we emit this diagnostic on the temporary object (which is in user code), so it is not suppressed.

I ran the diagnostic on the Google internal codebase, and I haven't found false positives so far. The only interesting case is [this issue](https://github.com/llvm/llvm-project/issues/97471), but that is a rare case where we should adjust the code. So, I believe there should not be many cases like this.

**Pull request merged**